cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+        }
+    }
+
+    /// Intersects each range with v.  If any intersection is empty, clears ranges.
+    void intersect_all(const query::clustering_range& v) {
+        for (auto& r : ranges) {
+            auto intrs = intersection(r, v, prefix3cmp);
+            if (!intrs) {
+                ranges.clear();
+                break;
+            }
+            r = *intrs;
+        }
+    }
+
+    template<std::ranges::range Range>
+    requires std::convertible_to<typename Range::value_type::value_type, bytes_opt>
+    void process_in_values(Range in_values) {
+        std::vector<query::clustering_range> new_ranges;
+        std::set<std::vector<bytes_opt>> tuples_seen;
+        for (const auto& current_tuple : in_values) {
+            if (tuples_seen.contains(current_tuple)) {
+                continue; // Storage proxy doesn't accept repeated identical ranges.
+            }
+            tuples_seen.insert(current_tuple);
+            // Each IN value is like a separate EQ restriction ANDed to the existing state.
+            auto current_range = to_range(
+                    oper_t::EQ, clustering_key_prefix::from_optional_exploded(*schema, current_tuple));
+            for (const auto& r : ranges) {
+                auto intrs = intersection(r, current_range, prefix3cmp);
+                if (intrs) {
+                    new_ranges.push_back(*intrs);
+                }
+            }
+        }
+        ranges = new_ranges;
+    }
+
+    static bytes deref_column_value(bytes_opt val, sstring_view name) {
+        return *statements::request_validations::check_not_null(
+                val, "Invalid null value in condition for column %s", name);
+    }
+};
+
+/// Calculates clustering bounds for the multi-column case.
+std::vector<query::clustering_range> get_multi_column_clustering_bounds(
+        const query_options& options,
+        schema_ptr schema,
+        const std::vector<expression>& multi_column_restrictions) {
+    multi_column_range_accumulator acc{options, schema};
+    for (const auto& restr : multi_column_restrictions) {
+        std::visit(acc, restr);
+    }
+    return acc.ranges;
+}
+
+/// For each element of a, creates b.size() new elements, each of which equals a's element extended by b's element.
+std::vector<std::vector<bytes>> accumulate_cartesian_product(
+        std::vector<std::vector<bytes>>&& a, const std::vector<bytes>& b) {
+    if (b.size() == 1) { // Don't need multiple copies of a's elements; just extend them in place.
+        for (auto& a_element : a) {
+            a_element.push_back(b[0]);
+        }
+        return move(a);
+    }
+    std::vector<std::vector<bytes>> product;
+    product.reserve(a.size() * b.size());
+    for (const auto& a_element : a) {
+        for (const auto& b_element : b) {
+            std::vector<bytes> extended = a_element;
+            extended.push_back(b_element);
+            product.push_back(extended);
+        }
+    }
+    return product;
+}
```

Comment:
Cartesian products can grow quite large. Did this lose the checks on cartesian product size we had before, or is it still checked?

Also, better return chunked_vector to avoid large continuous allocations.

Reply:
Ah, the check is below.

Reply:
Why are you doing a single multiplication at a time? IIRC the original code did all the multiplicands at once.

Reply:
> Also, better return chunked_vector to avoid large continuous allocations.

Will it really help, given that the end result must be std::vector due to the get_clustering_bounds() return type?

> Why are you doing a single multiplication at a time? IIRC the original code did all the multiplicands at once.

By "did" do you mean size checking or calculating the actual Cartesian product? I guess I should look into using the cartesian_product class...

Reply:
OK, I switched to using the existing cartesian_product utility, which should make us equivalent to the old code.  Specifically, it does "all the multiplicands at once".